### ISSUE TITLE
[FEATURE] New time series regression metrics

### ISSUE THREAD

**Issue by Garve** [koaning/scikit-lego#445]
Hi!

It's me again. How about some time series regression metrics which are not present in scikit-learn? So far, scikit-learn has the MAPE. This measure has some serious shortcomings, see [here](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error#Issues). If you are interested in this topic, the following could be good additions:

- Symmetric mean absolute percentage error
- Mean directional accuracy
- Mean Arctangent Absolute Percentage Error
- Logarithm of the accuracy ratio

If this is too much, let put some focus on the most interesting ones: Mean directional accuracy and the Logarithm of the accuracy ratio.

## Mean directional accuracy
I think this one is especially interesting when you only want to predict the direction of your time series correctly. Up or down nothing else. A classic problem to apply this is when dealing with the prediction of stock prices. I don't really care if my model can tell me if the stock price rises by 10% or 20%, it's important that it rises. And the model won't output accurate percentages anyway. It makes it potentially easier to design a well-performing model according to this metric.

## Logarithm of the accuracy ratio
[Here](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2635088), the author explains why it is a good measure. What I like about this measure: Usually, the MAPE punishes over and underestimating differently. If the actual value A is 2 and the forecast F is 1, the APE here is 0.5. If A=2 and F=4, the APE is 1. Two different scores, although the prediction is half or double the actual value, and I would like these errors to be the same. The Logarithm of the accuracy ratio does exactly that (0.48 in both cases). (By the way, the Mean directional accuracy does this, too.)

Another nice property is that the best constant prediction for some label vector y is the geometric mean of y. This is analog to the mean, which is the best constant prediction for the MSE metric, or the median for the MAD metric.

[You can find my implementations here.](https://github.com/Garve/scikit-bonus/blob/master/skbonus/metrics/_regression.py)

Best

Robert

**Comment by koaning** [comment]
I'm not opposed to the idea, but I'm wondering if there's scikit-learn timeseries packages that should be a better home for this sort of thing. You mention these metrics are specific to timeseries so did you check if these timeseries tools? Here's [one](https://github.com/alan-turing-institute/sktime).

**Comment by MBrouns** [comment]
As an aside, I never really got the "MAPE is not symmetric" issue, the wiki references [this paper](https://papers.ssrn.com/sol3/Delivery.cfm/SSRN_ID2635088_code402054.pdf?abstractid=2635088&mirid=1&type=2) which states

>The formula for MAPE is not symmetric in the sense that interchanging ŷ and y does not lead
to the same answer, despite the fact that the absolute error is the same before and after the
switch.

Isn't that exactly the point though? If absolute error mattered we would use a metric that measures absolute error. Predicting 100 if the true value was 130 is a smaller percentage difference than predicting 130 if the true value was 100? 

A fair comparison would be predicting 100 if the true value was 130 vs predicting 160 if the true value was 130 and to my knowledge both of those actually have the same mape

**Comment by Garve** [comment]
> I'm not opposed to the idea, but I'm wondering if there's scikit-learn timeseries packages that should be a better home for this sort of thing. You mention these metrics are specific to timeseries so did you check if these timeseries tools? Here's [one](https://github.com/alan-turing-institute/sktime).

Ah, true! It seems they don't have the MDA and LAR, though.

**Comment by Garve** [comment]
> As an aside, I never really got the "MAPE is not symmetric" issue, the wiki references [this paper](https://papers.ssrn.com/sol3/Delivery.cfm/SSRN_ID2635088_code402054.pdf?abstractid=2635088&mirid=1&type=2) which states
> 
> > The formula for MAPE is not symmetric in the sense that interchanging ŷ and y does not lead
> > to the same answer, despite the fact that the absolute error is the same before and after the
> > switch.
> 
> Isn't that exactly the point though? If absolute error mattered we would use a metric that measures absolute error. Predicting 100 if the true value was 130 is a smaller percentage difference than predicting 130 if the true value was 100?
> 
> A fair comparison would be predicting 100 if the true value was 130 vs predicting 160 if the true value was 130 and to my knowledge both of those actually have the same mape

That's true. If A is the actual value and F is the forecast, and let's say that F=pA for some percentage p. Then the APE of this term is |A-F|/|A| = |1-p|. If p is close to 0% (i.e. the model underestimates the data a lot), it yields an APE of around 100%, while a very large p such as 300% (the model overestimates a lot) gives an APE of 200%. And it is even unbounded for larger p.

The problem is the following: If you use the MAPE to compare models (such as using it as a scoring function while hyperparameter optimization), models that tend to underestimate have a lower MAPE, due to the boundedness of the APE in the case of underestimation.

**Comment by Garve** [comment]
> I'm not opposed to the idea, but I'm wondering if there's scikit-learn timeseries packages that should be a better home for this sort of thing. You mention these metrics are specific to timeseries so did you check if these timeseries tools? Here's [one](https://github.com/alan-turing-institute/sktime).

I presented it to the sktime project now.
